Initial runtime changes for composite support

trylek opened the pull request:
Together with the remaining counterpart Crossgen2 fixes, this change
seems good enough to let me pass a small sample of CoreCLR unit tests.
I am following up by broader testing over the weekend in the hope
that we can finalize the review and merge this in next week.

THanks

Tomas

trylek committed:
WIP: Initial CoreCLR runtime changes for composite R2R support

As of this initial commit the change is good enough so that,
upon initialization, it's able to load the composite R2R image
when opening System.Private.CoreLib and properly initialize it
including calls to the compiled methods in the composite image.
I'm also able to properly resolve import signatures in the
composite executable (i.o.w. I have already modified
DecodeModuleFromIndex).

I'm currently working on loading a secondary assembly which hits
interesting new issues like management of the
READYTORUN_HELPER_Module import cell in the composite case -
apparently we need different module cells for each component
assembly but we should be resolving them lazily, otherwise
opening the composite R2R image automatically forces load of
all component assemblies with potentially disastrous startup perf
effects.

Initially I'm hoping for 10000 feet feedback in the sense of
making sure I'm putting the changes in the correct places. Once
we agree on the basic concept, we can drill into the actual
implementation in more detail.

Thanks

Tomas

trylek committed:
Fix use of Windows-specific 'byte' type

trylek commented:
To sum up some of the fundamental design questions I've hit so far:

1) In this version I'm modelling the composite image as a new type of object that is not a Module. I'm starting to doubt whether that's reasonable, I'm thinking it might be easier to pretend it's a special type of Assembly / Module, mostly due to the way module cells are used in R2R helpers.

2) Right now there's no support for lifetime management of the composite images - my initial thinking is that these should be basically refcounted by the number of referencing composite MSIL assemblies - once this count gets to zero, the composite image should become unloadable.

3) I tried to follow some of the LoaderAllocator concepts but I'm not sure whether that's sufficient to make unloadability work end to end. Sadly JanV is on spring break vacation this week so I cannot consult him in this matter.

4) The way it works now is that multiple ReadyToRunInfo instances for the individual component assemblies point to separate per-assembly tables like entrypoints, available types or inlining info, while they share the image-wide tables like instance entrypoints, runtime functions, exception info or import tables. I'm not yet completely clear on whether this cannot have any adverse effect on the import cell resolution logic.

jkotas commented:
> In this version I'm modelling the composite image as a new type of object that is not a Module. 

I think it is a good idea. Note that we want to have an option to have native OS envelope (e.g. ELF) and then it having this to be Module would make even less sense.

>  lifetime management of the composite images

I think that the composite image should get loaded into AssemblyLoadContext and its lifetime should be tied to the lifetime of that AssemblyLoadContext.

> I tried to follow some of the LoaderAllocator concepts 

I do not think we have done the work to make R2R images unloadable.

trylek committed:
Fix for eager fixups & attribute presence table

Make sure we only run eager fixups once for the composite image;
fix attribute presence table to be properly per assembly.

trylek committed:
Fix reverse lookup of entry points to MethodDescs

trylek committed:
Fix member variable initialization order in constructor

lpereira commented on src/coreclr/src/utilcode/pedecoder.cpp:
```diff
@@ -2319,22 +2319,70 @@ READYTORUN_HEADER * PEDecoder::FindReadyToRunHeader() const
+    
+    // TODO: endianness?
+    // +0x18: number of name pointers
+    int namePointerCount = *(const int32_t *)&exportDirectory[0x18];
```
Are these bounds checked somewhere else?


Is `imageBase` supposed to be aligned on a int32_t?  Because if not, you might want to define a helper like this:

```C
static inline int32_t ReadLEInt32(const uint8_t *base, ptrdiff_t offset)
{
    int32_t value;

    memcpy(&value, base + offset, sizeof(value));
    return le32toh(value);
}
```

Which takes care of this not causing an unaligned load, and converts the little endian encoding to whatever the current CPU uses.

trylek replied:
I have added a simple helper ReadUInt32 to pedecoder.cpp to combine the range check, unaligned read and little endian decoding. Thanks for keeping an eye on robustness of our runtime!

lpereira commented on src/coreclr/src/utilcode/pedecoder.cpp:
```diff
@@ -2319,22 +2319,70 @@ READYTORUN_HEADER * PEDecoder::FindReadyToRunHeader() const
+
+    for (int nameIndex = 0; nameIndex < namePointerCount; nameIndex++)
+    {
+        int namePointerRVA = *(int32_t *)&imageBase[namePointersRVA + sizeof(int32_t) * nameIndex];
```
Is it possible that a corrupted PE file will make this go outside the bounds of imageBase?  Is it possible that the index expression might overflow?

trylek replied:
Fixed by means of using a new helper ReadUInt32 that includes the range check.

trylek committed:
Explicitly propagate PEImage when constructing DomainCompositeImage

Looks like this is the only reliable way to get to things like
the R2R header and such.

Thanks

Tomas

jkotas commented on src/coreclr/src/vm/appdomain.hpp:
```diff
@@ -1845,6 +1845,152 @@ class AppDomain : public BaseDomain
+    };  // class DomainAssemblyList
+
+    // Conceptually a list of DomainCompositeImage structures, protected by lock AppDomain::GetAssemblyListLock
+    DomainCompositeImageList m_CompositeImages;
```
It may be better for this list to live in AssemblyLoadContext (https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/assemblyloadcontext.h)

trylek replied:
Thanks for the suggestion, moved in 7th commit.

jkotas commented on src/coreclr/src/vm/domainfile.h:
```diff
@@ -755,7 +753,100 @@ class DomainAssembly : public DomainFile
+    static count_t Hash(const SString& a) { return a.Hash(); }
+};
+
+class DomainCompositeImage : public DomainFile
```
I am not sure whether inheriting `DomainCompositeImage` from `DomainFile` is the best design. It makes it have number of concepts that do not make sense for it. `DomainFile::GetExposedModuleObject`, `DomainFile::IsSystem`, the load levels, ...  none of it is applicable to the composite native image.

Would it be better if this is `class NativeImage` that does not inherit from anything?

trylek replied:
I have no problems with that whatsoever. I just had to start somewhere so I picked a place that seemed doing something remotely similar :-)… made the more interesting by the fact that JanV whom I'd normally consult on such matters first was OOF this week. Will move and refactor, sure.

trylek replied:
Refactored in 7th commit.

janvorli commented on src/coreclr/src/vm/domainfile.cpp:
```diff
@@ -2366,6 +2346,164 @@ DomainAssembly::EnumMemoryRegions(CLRDataEnumMemoryFlags flags)
+    return utf8Length == m_utf8SimpleNameLength && !memcmp(utf8SimpleName, m_utf8SimpleName, utf8Length);
+}
+
+bool DomainCompositeImage::TestAndClearRunEagerFixups()
```
The intent of this function is difficult to decipher from the name. How about something like "NeedRunEagerFixups" or something along those lines?

trylek replied:
For now I renamed this to <code>EagerFixupsNeedToRun</code>.

janvorli commented on src/coreclr/src/vm/domainfile.cpp:
```diff
@@ -2366,6 +2346,164 @@ DomainAssembly::EnumMemoryRegions(CLRDataEnumMemoryFlags flags)
+    }
+}
+
+DomainCompositeImage::~DomainCompositeImage()
```
Do we really need the empty destructor?

trylek replied:
I doubt so. Removed in 7th commit.

janvorli commented on src/coreclr/src/vm/domainfile.cpp:
```diff
@@ -2366,6 +2346,164 @@ DomainAssembly::EnumMemoryRegions(CLRDataEnumMemoryFlags flags)
+    const AssemblyNameIndex *assemblyNameIndex = m_assemblySimpleNameToIndexMap.Lookup(simpleName);
+    if (assemblyNameIndex != NULL)
+    {
+        const byte *imageBase = (const byte *)m_peImage->GetLoadedLayout()->GetBase();
```
I am surprised we even have a "byte" type. We use "BYTE" all over the place, so I'd stick with that.

trylek replied:
Thanks for spotting, I hit this before on one of my other change as the type only exists on Windows. Renamed in 7th commit.

janvorli commented on src/coreclr/src/vm/readytoruninfo.h:
```diff
@@ -48,13 +74,18 @@ class ReadyToRunInfo
     static PTR_ReadyToRunInfo Initialize(Module * pModule, AllocMemTracker *pamTracker);
 
+    bool IsComponentAssembly() const { return m_isComponentAssembly; }
+    const ReadyToRunCoreInfo * GetComponent() const { return &m_component; }
```
A nit - it seems that GetComponentInfo / GetCompositeInfo or something along those lines would be a better to understand name.

janvorli commented on src/coreclr/src/vm/readytoruninfo.h:
```diff
@@ -20,15 +20,40 @@ typedef DPTR(struct READYTORUN_SECTION) PTR_READYTORUN_SECTION;
+    ReadyToRunInfo*                 m_compositeInfo;
+
+    ReadyToRunCoreInfo              m_component;
+    const ReadyToRunCoreInfo*       m_composite;
```
A nit - in the runtime, we tend to prefix pointer variable names with `p`, so for example this one would be m_pComposite.

trylek replied:
Will do - I was originally under the impression that this is generally a remnant of the no longer used Hungarian notation.

janvorli commented on src/coreclr/src/vm/zapsig.cpp:
```diff
@@ -677,16 +684,19 @@ Module *ZapSig::DecodeModuleFromIndexIfLoaded(Module *fromModule,
     {
-        if (index < fromModule->GetAssemblyRefMax())
+        DomainCompositeImage *compositeImage = fromModule->GetCompositeImage();
+        uint32_t assemblyRefMax = (compositeImage != NULL ? 1 : fromModule->GetAssemblyRefMax());
```
Should this rather get 0 for composite images?

trylek replied:
Fixed. This requires bits of counterpart changes to Crossgen2 and R2RDump I am adding to the PR in 7th commit.

trylek commented on src/coreclr/src/inc/readytorun.h:
```diff
@@ -62,6 +75,8 @@ enum class ReadyToRunSectionType : uint32_t
     ManifestMetadata            = 112, // Added in V2.3
     AttributePresence           = 113, // Added in V3.1
     InliningInfo2               = 114, // Added in V4.1
+    ComponentAssemblies         = 115, // Added in V4.1
```
I think that 4.1 hasn't shipped yet so it should be fine. @jkotas, could you please confirm?

jkotas replied:
Right

trylek commented on src/coreclr/src/vm/appdomain.cpp:
```diff
@@ -2684,6 +2684,7 @@ AppDomain::AppDomain()
 
 #ifdef _DEBUG
     m_Assemblies.Debug_SetAppDomain(this);
+    m_CompositeImages.Debug_SetAppDomain(this);
```
I have reverted most of this code as part of moving the native image list and related logic from appdomain to assemblyloadcontext. I can easily do the cleanup you're suggesting but I'd prefer to do that as a separate change as it's no longer related to this change.

trylek commented on src/coreclr/src/vm/appdomain.cpp:
```diff
@@ -3516,6 +3517,62 @@ CHECK AppDomain::CheckCanExecuteManagedCode(MethodDesc* pMD)
+    if (compositeImage != NULL)
+    {
+        CrstHolder ch(GetAssemblyListLock());
+        m_CompositeImages.Append_Unlocked(compositeImage);
```
As part of the move to assemblyloadcontext I have for now removed the assembly list lock holders and I only used the LoadLockHolder to protect the entire composite image load. I'm not sure if this is sufficient as I'm not super familiar with general CoreCLR locking strategies. I'll welcome any suggestions as to what is the proper way to protect the native image load.

trylek commented on src/coreclr/src/vm/appdomain.hpp:
```diff
@@ -1845,6 +1845,152 @@ class AppDomain : public BaseDomain
+            return Set_Unlocked(index, pCompositeImage);
+        }
+
+        void Set_Unlocked(DWORD index, DomainCompositeImage* pCompositeImage)
```
For now I have only kept the unlocked variants of all the methods. I plan to clean this up and unify once I understand what is the appropriate locking scheme for the native images.

trylek committed:
Address PR feedback from both Jan's and from Leandro

I have taken the DomainCompositeImage out of AppDomain and I
put it into AssemblyLoadContext per JanK's suggestion. I believe
I have addressed all feedback from JanV and I added a helper
for checked little endian 32-bit read from PE executables per
Leandro's suggestion.

Thanks

Tomas

trylek commented:
I believe I have addressed all PR feedback. Please take another look when you have a chance.

jkotas commented:
The crossgen tests are failing...

trylek committed:
Fix fallout from the module indexing refactoring

jkotas commented on src/coreclr/src/utilcode/pedecoder.cpp:
```diff
@@ -2319,22 +2319,95 @@ READYTORUN_HEADER * PEDecoder::FindReadyToRunHeader() const
+{
+    outValue = 0;
+
+    if (offset < 0 || (uint32_t)offset + sizeof(int) > GetSize())
```
I do not think you need to do bounds checks like this here. They do not actually guarantee that you won't crash.

I would just cast the pointer to `IMAGE_EXPORT_DIRECTORY` and be done with it.

trylek replied:
Simplified based on your suggestion in 9th commit.

jkotas commented on src/coreclr/src/utilcode/pedecoder.cpp:
```diff
@@ -2319,22 +2319,95 @@ READYTORUN_HEADER * PEDecoder::FindReadyToRunHeader() const
+
+#ifndef DACCESS_COMPILE
+    const BYTE *base = (const BYTE *)GetBase() + offset;
+    outValue = (base[0] << 0) | (base[1] << 8) | (base[2] << 16) | (base[3] << 24);
```
`VAL32` macro is used to deal with endianess everywhere else in this file, so just use that. (Not that we care about bigendian for CoreCLR these days...)

trylek replied:
Fixed in 9th commit.

jkotas commented on src/coreclr/src/vm/assemblyloadcontext.h:
```diff
@@ -15,6 +20,30 @@ class AssemblyLoadContext : public IUnknownCommon<ICLRPrivBinder, IID_ICLRPrivBi
+
+private:
+    // Multi-thread safe access to the list of composite R2R images
+    class NativeImageList
```
You can just use `SArray<NativeImage*>` instead of this.

trylek replied:
Fixed in 10th commit.

jkotas commented on src/coreclr/src/vm/assemblyloadcontext.h:
```diff
@@ -15,6 +20,30 @@ class AssemblyLoadContext : public IUnknownCommon<ICLRPrivBinder, IID_ICLRPrivBi
+    NativeImage *LoadNativeImage(Module *componentModule, LPCUTF8 nativeImageName, int nativeImageNameLength);
+
+private:
+    // Multi-thread safe access to the list of composite R2R images
```
Nit: This structure does not appear to be multi-thread safe on its own.

trylek replied:
Removed obsolete comment in 9th commit.

jkotas commented on src/coreclr/src/vm/common.h:
```diff
@@ -123,6 +123,7 @@ typedef DPTR(class ComPlusCallMethodDesc) PTR_ComPlusCallMethodDesc;
 typedef VPTR(class DebugInterface)      PTR_DebugInterface;
 typedef DPTR(class Dictionary)          PTR_Dictionary;
 typedef VPTR(class DomainAssembly)      PTR_DomainAssembly;
+typedef VPTR(class DomainCompositeImage) PTR_DomainCompositeImage;
```
This should not be needed anymore

trylek replied:
Good point, removed in 10th commit.

jkotas commented on src/coreclr/src/vm/nativeimage.h:
```diff
@@ -0,0 +1,87 @@
+class PEFile;
+class PEImage;
+
+class NativeImage
```
A comment explaining what this is representing may be useful

trylek replied:
Added explanatory comment in 9th commit.

trylek commented on src/coreclr/src/vm/assemblyloadcontext.cpp:
```diff
@@ -14,3 +15,130 @@ HRESULT AssemblyLoadContext::GetBinderID(
+{
+    CONTRACTL
+    {
+        if (FORBIDGC_LOADER_USE_ENABLED()) NOTHROW; else THROWS;
```
Fixed in 8th commit.

trylek commented on src/coreclr/src/vm/assemblyloadcontext.cpp:
```diff
@@ -14,3 +15,130 @@ HRESULT AssemblyLoadContext::GetBinderID(
     return S_OK;
 }
+
+NativeImage *AssemblyLoadContext::LoadNativeImage(Module *componentModule, LPCUTF8 nativeImageName, int nativeImageNameLength)
```
Fixed in 9th commit.

trylek committed:
Address JanK's PR feedback

trylek committed:
Address remaining JanK's feedback

jkotas commented on src/coreclr/src/vm/nativeimage.cpp:
```diff
@@ -0,0 +1,125 @@
+    }
+    LoaderHeap *pHeap = pLoaderAllocator->GetHighFrequencyHeap();
+    AllocMemTracker amTracker;
+    NativeImage *result = new (amTracker.Track(pHeap->AllocMem((S_SIZE_T)sizeof(NativeImage))))
```
We typically use loader allocator to only allocate types that only contain other items allocated by loader allocator. It is not the case here: NativeImage has multiple items that are allocated on regular heap. I think it would be better to allocated NativeImage on regular heap.

davidwrighton replied:
To be precise, the rule is, never allocate via a LoaderHeap a C++ object that has an implicit or explicit destructor. NativeImage has an implicit destructor due to the presence of a NewHolder and an SHash.

---
In reply to: [384631579](https://github.com/dotnet/runtime/pull/32500#discussion_r384631579) [](ancestors = 384631579)

trylek replied:
Fixed in 11th commit.

jkotas commented on src/coreclr/src/vm/assemblyloadcontext.cpp:
```diff
@@ -14,3 +15,49 @@ HRESULT AssemblyLoadContext::GetBinderID(
+    for (int nativeImageIndex = 0; nativeImageIndex < nativeImageCount; nativeImageIndex++)
+    {
+        NativeImage *nativeImage = m_nativeImages[nativeImageIndex];
+        if (nativeImage->Matches(nativeImageName, loadContext))
```
We should only have native images from the right load context here, so it should not be necessary to test loadContext for the match.

trylek replied:
Good point, removed.

trylek commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/OwnerCompositeExecutableNode.cs:
```diff
@@ -40,6 +40,7 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
             builder.RequireInitialPointerAlignment();
```
Good point, removed.

davidwrighton commented on src/coreclr/src/vm/readytoruninfo.h:
```diff
@@ -48,13 +75,18 @@ class ReadyToRunInfo
 
+    bool IsComponentAssembly() const { return m_isComponentAssembly; }
+    const ReadyToRunCoreInfo * GetComponentInfo() const { return &m_component; }
+    const ReadyToRunCoreInfo * GetCompositeInfo() const { return m_pComposite; }
```
This function isn't used.

trylek replied:
Refactored per your other suggestion in 11th commit.

davidwrighton commented on src/coreclr/src/vm/readytoruninfo.h:
```diff
@@ -48,13 +75,18 @@ class ReadyToRunInfo
+    bool IsComponentAssembly() const { return m_isComponentAssembly; }
+    const ReadyToRunCoreInfo * GetComponentInfo() const { return &m_component; }
+    const ReadyToRunCoreInfo * GetCompositeInfo() const { return m_pComposite; }
+    NativeImage * GetNativeImage() const { return m_pNativeImage; }
```
The vast majority of the uses of this function are GetCompositeInfo()->GetImage(). I feel that there should be just a GetImage() function on this type instead to reduce the verbosity, and hold the invariant that ReadyToRunInfo can only come from one file.

Also, given that that leaves only 2-3 uses of GetCompositeInfo() left, I think it might be worth doing the refactoring to entirely hide away the concept of composite vs component from consumers of this class.

trylek replied:
I love this suggestion, fixed in 11th commit.

trylek commented on src/coreclr/src/utilcode/pedecoder.cpp:
```diff
@@ -2319,22 +2319,66 @@ READYTORUN_HEADER * PEDecoder::FindReadyToRunHeader() const
+    }
+    
+    const uint8_t *imageBase = (const uint8_t *)GetBase();
+    const uint8_t *exportDirectory = imageBase + VAL32(exportDirectoryEntry->VirtualAddress);
```
Fixed in 11th commit.

trylek commented on src/coreclr/src/utilcode/pedecoder.cpp:
```diff
@@ -2319,22 +2319,66 @@ READYTORUN_HEADER * PEDecoder::FindReadyToRunHeader() const
         }
     }
+    
+    READYTORUN_HEADER *nativeReadyToRunHeader = FindNativeReadyToRunHeader();
```
Changed in 11th commit per your suggestion.

trylek commented on src/coreclr/src/vm/nativeimage.h:
```diff
@@ -0,0 +1,95 @@
+class PEFile;
+class PEImage;
+
+/// <summary>
```
Changed to normal double slashes. I was originally under the impression that XML comments are a general VS feature not limited to C# but I don't insist.

trylek commented on src/coreclr/src/vm/nativeimage.h:
```diff
@@ -0,0 +1,95 @@
+class PEImage;
+
+/// <summary>
+/// This class represents a native ReadyToRun image. As of today, this file format is used
```
Changed in 11th commit.

trylek commented on src/coreclr/src/vm/nativeimage.cpp:
```diff
@@ -0,0 +1,125 @@
+
+bool NativeImage::Matches(LPCUTF8 utf8SimpleName, const AssemblyLoadContext *pLoadContext) const
+{
+    return pLoadContext == m_pLoadContext && !strcmp(utf8SimpleName, m_utf8SimpleName);
```
Changed to stricmp in 11th commit.

trylek commented on src/coreclr/src/vm/nativeimage.h:
```diff
@@ -0,0 +1,95 @@
+{
+private:
+    AssemblyLoadContext *m_pLoadContext;
+    LPCUTF8 m_utf8SimpleName;
```
It's just a pointer into the component MSIL image (into its OwnerCompositeExecutable R2R header section).

jkotas replied:
Ok, I see it is a pointer into MSIL image that happened to be the first one to load it. It would deserve a comment at least...

trylek replied:
I have added an explanatory comment in 13th commit.

davidwrighton commented on src/coreclr/src/vm/zapsig.cpp:
```diff
@@ -625,34 +626,41 @@ Module *ZapSig::DecodeModuleFromIndex(Module *fromModule,
+            {
+                pAssembly = nativeImage->LoadComponentAssembly(index);
+            }
+            else
```
Why add a function here, but not in DecodeModuleFromIndexIfLoaded?

trylek replied:
Ouch, thanks for pointing that out. I missed this second method when applying the module index refactoring suggested by JanV.

trylek replied:
Fixed in 11th commit.

davidwrighton commented on src/coreclr/src/vm/zapsig.cpp:
```diff
@@ -677,16 +685,19 @@ Module *ZapSig::DecodeModuleFromIndexIfLoaded(Module *fromModule,
     {
-        if (index < fromModule->GetAssemblyRefMax())
+        NativeImage *compositeImage = fromModule->GetCompositeNativeImage();
+        uint32_t assemblyRefMax = (compositeImage != NULL ? 1 : fromModule->GetAssemblyRefMax());
```
>1 [](start = 60, length = 1)

This isn't the same as the algorithm above in DecodeModuleFromIndex. (And this path is, I believe very rarely taken for R2R, so testing to success here is likely to be hard.)

trylek replied:
That's exactly why I missed that. I have modified it to use basically the same code as DecodeModuleFromIndex.

trylek replied:
Fixed in 11th commit.

trylek commented on src/coreclr/src/vm/nativeimage.h:
```diff
@@ -0,0 +1,95 @@
+    
+    NewHolder<ReadyToRunInfo> m_pReadyToRunInfo;
+    IMDInternalImport *m_pManifestMetadata;
+    PEImage *m_pPeImage;
```
It looks like the only reason for PEImage is GetNativeMDImport. Sadly it looks like this is why we also need the R2R header detection logic in pedecoder as the manifest metadata decoder depends on the presence of the header. Please let me know what are your thoughts towards disentangling this - I guess most of the craziness stems from the fact that Andon reused the NGen native metadata concept for this purpose.

jkotas replied:
It is trivial to open the metadata from a blob. Is it possible to just do that in NativeImage constructor instead of going round about way through PEImage to do it?

The use of manifest metadata in zapsig decoder is special cases already, so I assume that one should be fine.

trylek replied:
Fixed in 12th commit.

trylek commented on src/coreclr/src/vm/nativeimage.cpp:
```diff
@@ -0,0 +1,125 @@
+
+bool NativeImage::EagerFixupsNeedToRun()
+{
+    bool runEagerFixups = m_runEagerFixups;
```
You're right - I originally tricked myself into thinking this is protected by the native image load lock but the eager fixups take place at a later stage and can be called concurrently from multiple modules. I have added a locking scheme to support this in 11th commit.

trylek committed:
Address most of JanK's and DavidWr's PR feedback

The only feedback I skipped for now is removing the dual detection
of R2R header from pedecoder as it's currently needed for opening
the manifest metadata in the composite R2R image. I'll be happy to
clean this up once we figure out what's the ideal way to shape this.

Thanks

Tomas

trylek commented:
OK, I have finally pushed the 11th commit after making sure the change still works. I believe I have adressed most of PR feedback from @davidwrighton and @jkotas except for JanK's suggestion to remove the dual R2R header detection from pedecoder as it's currently needed for the manifest metadata in the composite R2R executable. I'll welcome any suggestions as to how to clean this up, ideally as a follow-up change to unblock downstream efforts related to composite R2R.

trylek committed:
Refactor manifest metadata processing per JanK's feedback

jkotas commented on src/coreclr/src/vm/nativeimage.cpp:
```diff
@@ -0,0 +1,137 @@
+    m_componentAssemblyCount = m_pComponentAssemblies->Size / sizeof(READYTORUN_COMPONENT_ASSEMBLIES_ENTRY);
+    
+    // Check if the current module's image has native manifest metadata, otherwise the current->GetNativeAssemblyImport() asserts.
+    m_pManifestMetadata = LoadManifestMetadata();
```
Add a `NativeImage::~NativeImage` destructor and release `m_pManifestMetadata` in it.

trylek replied:
Fixed in 13th commit.

jkotas commented on src/coreclr/src/vm/nativeimage.cpp:
```diff
@@ -0,0 +1,125 @@
+    {
+        LPCSTR assemblyName;
+        hr = m_pManifestMetadata->GetAssemblyRefProps(assemblyRef, NULL, NULL, &assemblyName, NULL, NULL, NULL, NULL);
+        m_assemblySimpleNameToIndexMap.Add(new (amTracker.Track(pHeap->AllocMem((S_SIZE_T)sizeof(AssemblyNameIndex))))
```
Do you plan to address this one?

jkotas replied:
The traits for the table should be something like this:
```
struct AssemblyNameIndex
{
    LPCUTF8 Name;
    int32_t Index;
};


class AssemblyNameIndexHashTraits : public NoRemoveSHashTraits< DefaultSHashTraits<AssemblyNameIndex> >
{
public:
    typedef LPCUTF8 key_t;

    static key_t GetKey(element_t e)
    {
        return e.Name;
    }
    static BOOL Equals(key_t k1, key_t k2)
    {
        return ....;
    }
    static count_t Hash(key_t k)
    {
        return ....;
    }
    static const element_t Null()
    {
        AssemblyNameIndex e;
        e.Name = NULL;
        e.Index = 0;
        return e;
    }
    static bool IsNull(const element_t &e)
    {
        return e.Name == NULL;
    }
};
```

jkotas commented on src/coreclr/src/vm/assemblyloadcontext.cpp:
```diff
@@ -14,3 +15,49 @@ HRESULT AssemblyLoadContext::GetBinderID(
+    fullPath += compositeImageFileName;
+
+    PTR_PEImage peImage = PEImage::OpenImage(fullPath);
+    PTR_PEFile peFile = PEFile::Open(peImage);
```
You should not need to open PEFile or PEImage. Just create PEImageLayout without the extra ceremony.

trylek replied:
I apologize, I'm not getting this - how can I get a PEImageLayout without a PEImage?

jkotas replied:
For Windows, it can look like this:

```
#ifdef TARGET_WINDOWS
    m_hModule = CLRLoadLibraryEx(path,NULL,GetLoadWithAlteredSearchPathFlag());
    if (m_hModule == NULL)
    {
        HRESULT hr = HRESULT_FROM_GetLastError();
        EEFileLoadException::Throw(path, hr, NULL);
    }
    m_pPEDecoder = new PEDecoder((void*)m_hModule, TRUE);
#else
```

For Unix, it will be a bit more complex - the key functions to call are `PAL_LOADLoadPEFile` and `ApplyBaseRelocations()`.

The reason to avoid PEImage is that it has multiple ways to load stuff, and it does fallbacks between different ways for compatibility. You do not need any of that for this: You know that the file should be a good native image, and if it is not - the right thing to do is to fail immediately.

trylek replied:
Thanks for the explanation. I'm trying to make this change but I guess I'm still missing how to construct a PEImageLayout based on the snippet you kindly provided and switching over all related logic to use PEDecoder seems pretty viral, especially due to <code>Module::GetNativeOrReadyToRunImage</code>. If that is what you're truly suggesting, I'll probably make a preparatory change just "downgrading" these various places to use PEDecoder instead of PEImageLayout to simplify integration of this change.

jkotas replied:
Keeping PEImageLayout is ok - I have not realized that it is viral. PEImage and PEFile are the important ones to avoid.

trylek committed:
More JanK's PR feedback

trylek commented:
@jkotas - OK, the only two remaining open issues I'm aware of: 1) simplifying PEFile -> PEImage -> PEImageLayout code flow you seem to be alluding to in your feedback; 2) I'm still not getting how to refactor the assembly index map in NativeImage to use the string pointers as the keys - is there any precedent I can look at for inspiration?

davidwrighton commented on src/coreclr/src/vm/assemblyloadcontext.cpp:
```diff
@@ -14,3 +15,49 @@ HRESULT AssemblyLoadContext::GetBinderID(
+        m_nativeImages.Append(nativeImage);
+        return nativeImage;
+    }
+#endif
```
All of this logic feels temporary until we integrate all of this with the hosting surface and such. From my understanding, you've been putting together a system where we place multiple IL "PE" images into a single large file, and this is getting at the PE image of the actual wrapping large file. I don't understand why this needs to disk, instead of pulling the PEImage layout which is already loaded, somewhere. At the very least, this method needs a comment describing its purpose and description of how it attaches to the load flow of the runtime.

trylek replied:
Hmm, I'm not sure if I understand your comment. At this point I'm not trying to put "multiple IL PE images into a single large file" - as suggested by JanK and described in the readytorun-format.md design doc, I'm currently working on implementing "composite R2R with standalone MSIL" which means that a number of MSIL assemblies (the app, the framework, whatever) continues to reside in individual files on disk, they just share the "native code cache" represented by the compiled methods in the composite R2R file. I can easily improve the code comments or other documentation but I guess I first need to make sure the design is communicated clearly enough so that it makes sense at least to experts like you.

davidwrighton replied:
Well, my comment is that in the long run IMO, there should not be *any* logic that is aware of file paths in the core engine of the runtime. Obviously as you put together the feature you have found that you need something here, but we will need a plan to remove all of that to an external, configurable bit of logic like the host.

As for the details around which variant of composite image you are making, the logic as written is extremely similar to what I would expect for the composite image model with embedded MSIL PE files. Its just lacking a discovery process for the embedded MSIL concept. I simply missed the distinction on production. (I blame going on vacation for a week.)

trylek replied:
Thanks for clarification, that makes sense to me. The decision to start with standalone MSIL is purely pragmatic as that requires fewer changes in the runtime exactly due to the fact that no changes are needed to MSIL discovery. I don't yet have a very clear idea as to whether the support for composite R2R with standalone MSIL is purely temporary or will remain as a longer-term supported scenario, I agree that the owner composite executable is a bit of an opportunistic hack rather than something backed by long-term strategic thinking.

jkotas commented on src/coreclr/src/vm/nativeimage.cpp:
```diff
@@ -0,0 +1,150 @@
+    m_pPeImageLayout = pPeImageLayout;
+    m_eagerFixupsHaveRun = false;
+    
+    m_pReadyToRunInfo.Assign(new (amTracker.Track(pHeap->AllocMem((S_SIZE_T)sizeof(ReadyToRunInfo))))
```
This is allocated on LoaderHeap, but stored in NewHolder. NewHolder is for things allocated using regular `new`. Things allocated on LoaderHeap should not be stored in NewHolder.

It may be best to use regular new for everything related to this, and not use LoaderHeap at all.

trylek replied:
Fixed in 14th commit.

jkotas commented on src/coreclr/src/vm/nativeimage.cpp:
```diff
@@ -0,0 +1,150 @@
+    READYTORUN_HEADER *pHeader = (READYTORUN_HEADER *)((BYTE *)pPeImageLayout->GetBase() + headerRVA);
+    AllocMemTracker amTracker;
+    NativeImage *result = new NativeImage(pPeFile, pPeImageLayout, pHeader, nativeImageName, pLoaderAllocator, amTracker);
+    amTracker.SuppressRelease();
```
If you would like to keep using LoaderHeap, SuppressRelease should be done only after the structure is registered with global state - that would be after `m_nativeImages.Append(nativeImage);`.

trylek replied:
Hmm, now I'm somewhat confused - ReadyToRunInfo ctor does need an AllocMemTracker argument and the "legacy" ReadyToRunInfo::Initialize does initialize the ReadyToRunInfo using the LoaderHeap and AllocMemTracker. I was originally trying to follow the same pattern but if it's not the right pattern to use, I could probably use some advice as to what is the preferable pattern here.

jkotas replied:
Ah ok. It is fine to keep allocating some of this on LoaderHeap with AllocMemTracker to follow existing pattern.

trylek replied:
Moved AllocMemTracker to AssemblyLoadContext::LoadNativeImage in 14th commit per your suggestion.

jkotas commented on src/coreclr/src/utilcode/pedecoder.cpp:
```diff
@@ -2335,6 +2335,41 @@ READYTORUN_HEADER * PEDecoder::FindReadyToRunHeader() const
 
+uint32_t PEDecoder::GetExport(LPCSTR exportName) const
+{
+#ifndef DACCESS_COMPILE
```
Ifdefing the body of the method with `DACCESS_COMPILE` makes it do a wrong thing for DAC build. It is not a desirable pattern that has been source of many subtle bugs.

- Make the method compatible with DAC and have no ifdefs
- Or, ifdef out the whole method so that you can be sure that it is not used under DAC
- Or, make the method body call `DacNotImpl();`. This one is still not ideal because of the method is still not going to work when called under DAC, but at least the failure is easy to see and diagnose.

trylek replied:
Thanks for pointing that out, fixed in 14th commit.

trylek committed:
Address remaining JanK's PR feedback

1) Removed conditional behavior based on DACCESS_COMPILE (all
remaining conditional compilation should have method granularity);

2) Removed use of PEFile and PEImage and changed NativeImage to
directly use PEImageLayout per JanK's suggestions;

3) Fixed case insensitive assembly name comparison the same way
BaseAssemblySpec does it - by using temporary SStrings. It's
certainly not too efficient but it shouldn't be on a hot path
(the number of times the hashmap lookup is performed equals the
number of loaded assemblies using composite images).

Thanks

Tomas

trylek committed:
Fix Linux build

jkotas commented on src/coreclr/src/inc/pedecoder.h:
```diff
@@ -402,7 +407,7 @@ class PEDecoder
     CORCOMPILE_HEADER *FindNativeHeader() const;
     READYTORUN_HEADER *FindReadyToRunHeader() const;
-
+    
```
Nit: Extra whitespace

trylek replied:
Thanks for spotting, fixed in 16th commit.

jkotas commented on src/coreclr/src/vm/nativeimage.h:
```diff
@@ -0,0 +1,108 @@
+{
+private:
+    // Points to the OwnerCompositeExecutable section content within the component MSIL module
+    LPCUTF8 m_utf8SimpleName;
```
Is this a simple name (ie binary name without any prefix/suffix) or is this actually a filename? If I am reading the code right, it should be a filename.

trylek replied:
You're right, it is supposed to include the extension. Renamed to m_utf8FileName in 16th commit.

jkotas replied:
I would just call it `m_pFileName`. I do not think we have `m_utf8XXX` anywhere else.

jkotas commented on src/coreclr/src/vm/nativeimage.cpp:
```diff
@@ -0,0 +1,156 @@
+
+bool NativeImage::Matches(LPCUTF8 utf8SimpleName) const
+{
+    return !_stricmp(utf8SimpleName, m_utf8SimpleName);
```
It this is actually a filename, it would be more appropriate to conditionalize this using `FEATURE_CASE_SENSITIVE_FILESYSTEM` like we do in other places.

trylek replied:
Good point, fixed in 16th commit.

lpereira replied:
The other place (`AssemblyNameIndexHashTraits::Equals()`) that does a case-insensitive string comparison of LPCUTC8 values converts this to SStrings and whatnot; maybe just use _stricmp() like here and avoid creating those temporaries altogether?

trylek replied:
Thanks for pointing that out. That was actually a bug, I believe <code>_stricmp</code> isn't guaranteed to perform a correct UTF8 case insensitive comparison in the general case even though the way Unicode and UTF8 are constructed it may just work in common cases. I have replaced this bit to use <code>SString</code>-based case insensitive comparison too.

jkotas commented on src/coreclr/src/vm/nativeimage.cpp:
```diff
@@ -0,0 +1,156 @@
+    }
+}
+
+bool NativeImage::Matches(LPCUTF8 utf8SimpleName) const
```
Nit: It may be better to have a method that returns the NativeImage name instead, and do the comparison in the caller instead.

trylek replied:
Fixed in 16th commit.

jkotas commented on src/coreclr/src/vm/nativeimage.cpp:
```diff
@@ -0,0 +1,156 @@
+{
+    WRAPPER_NO_CONTRACT;
+
+    SString s1;
```
`return SString(SString::Utf8, a).CompareCaseInsensitive(SString(SString::Utf8, b));`

jkotas replied:
Actually, it can even be:
`return SString(SString::Utf8Literal, a).CompareCaseInsensitive(SString(SString::Utf8Literal, b));`

(saves unnecessary copy)

trylek replied:
Thanks for pointing that out, fixed in 16th commit.

jkotas commented on src/coreclr/src/vm/nativeimage.cpp:
```diff
@@ -0,0 +1,156 @@
+    NewHolder<PEImageLayout> peLoadedImage = PEImageLayout::LoadNative(fullPath);
+
+    uint32_t headerRVA = peLoadedImage->GetExport("RTR_HEADER");
+    if (headerRVA + sizeof(READYTORUN_HEADER) > peLoadedImage->GetSize())
```
This should not be a silent failure. We should throw exception here.

jkotas replied:
Instead of sizes, it would be better to check R2R Signature and Version

lpereira replied:
Is there a possibility that the `headerRVA` value, from a corrupted/maliciously-crafted binary, could be overflown when adding it to `sizeof(READYTORUN_HEADER)` in such a way that this check is bypassed?

Usually, rewriting such checks prioritizing compile-time constants and doing subtractions instead of additions will often suffice:

```C
    if (peLoadedImage->GetSize() - sizeof(READYTORUN_HEADER) <= headerRVA)
    {
          ...
    }
```

jkotas replied:
If you get corrupted or maliciously crafted binary here, you can cause arbitrary or anything else.

This should focus on high-value sanity checks that filter out user mistakes. Checking the R2R signature and Version is example of high-value sanity check.

trylek replied:
Added signature and version check throwing a failure on mismatch in 16th commit.

jkotas commented on src/coreclr/src/inc/pedecoder.h:
```diff
@@ -349,6 +351,9 @@ class PEDecoder
     void *GetNativeEntryPoint() const;
 
+    // Look up a named symbol in the export directory
+    uint32_t GetExport(LPCSTR exportName) const;
```
Why not return the actual pointer to the export here? (And null if it does not exist.)

trylek replied:
Good point, changed in 16ht commit.

lpereira commented on src/coreclr/src/utilcode/pedecoder.cpp:
```diff
@@ -2335,6 +2335,41 @@ READYTORUN_HEADER * PEDecoder::FindReadyToRunHeader() const
+
+    for (uint32_t nameIndex = 0; nameIndex < namePointerCount; nameIndex++)
+    {
+        uint32_t namePointerRVA = VAL32(*(const uint32_t *)&imageBase[namePointersRVA + sizeof(uint32_t) * nameIndex]);
```
I'm still concerned that these will be an out of bounds access for corrupted/maliciously-crafted binaries.

trylek replied:
I have simplified this code based on explicit Jan's request. Without being completely familiar or trying to interpret the official view, I think that the rule of thumb is that making sure binaries haven't been tampered with is not a matter for this low-level runtime, that's a question of deployment options, signatures etc. There are cases where even a legal PE file can crash the runtime, e.g. upon random disk sector failure. I defer to @jkotas for additional thoughts in this matter.

jkotas commented on src/coreclr/src/vm/nativeimage.cpp:
```diff
@@ -0,0 +1,156 @@
+    LPCUTF8 nativeImageName,
+    LoaderAllocator *pLoaderAllocator,
+    AllocMemTracker *pamTracker)
+{
```
It may be nice to have `STANDARD_VM_CONTRACT;` on all these methods for clarity and to validate that they do not have any weird interations with GC, etc.

trylek replied:
Added to all methods in 16th commit even though I'm not sure if it's needed on the destructor.

jkotas commented on src/coreclr/src/vm/nativeimage.h:
```diff
@@ -0,0 +1,108 @@
+    // Points to the OwnerCompositeExecutable section content within the component MSIL module
+    LPCUTF8 m_utf8SimpleName;
+    
+    NewHolder<ReadyToRunInfo> m_pReadyToRunInfo;
```
The CoreCLR codebase typically does not use `NewHolder` as field. It has an extra overhead and it is not thread safe. The typical pattern is to store the raw pointer and release it in the destructor explicitly.

trylek replied:
Fixed in 16th commit.

jkotas commented on src/coreclr/src/vm/nativeimage.cpp:
```diff
@@ -0,0 +1,156 @@
+}
+
+NativeImage::NativeImage(
+    NewHolder<PEImageLayout>& peImageLayout,
```
It is unusual to pass NewHolder by reference like this in CoreCLR codebase. The usual pattern is to have a constructor that takes ownership and cannot fail half-way through, and Initialize method that does the initialization that can fail in arbitrary ways.

trylek replied:
Fixed in 16th commit.

jkotas reviewed (approved):
LGTM modulo last set of nits. Thank you!

trylek committed:
Address remaining Jan's nits

trylek committed:
Fix library propagation in SuperIlc on Linux

janvorli commented on src/coreclr/src/tools/ReadyToRun.SuperIlc/BuildFolder.cs:
```diff
@@ -21,12 +21,16 @@ public class BuildFolder
 
+        private static string[] s_runtimeWindowsOnlyLibraries =
+        {
+            "mscorrc",
```
This has been just modified by #33123, we don't have the mscorrc.debug anymore.

trylek replied:
Beautiful, thanks for pointing that out! Removed in 18th commit.

trylek commented:
Merging in with one pre-existing Crossgen2 failure, tracked under https://github.com/dotnet/runtime/issues/33155.

jkotas commented:
This broke debugging because of the code is not DACized correctly. 

windbg is enable to resolve managed symbols in R2R images, etc. This makes us unable to triage and diagnose other crashes happening on CI. I am going to revert this.

cc @dotnet/runtime-infrastructure @dotnet/dotnet-diag

jkotas commented on src/coreclr/src/vm/readytoruninfo.h:
```diff
@@ -18,17 +18,43 @@
-    PTR_PEImageLayout               m_pLayout;
     PTR_READYTORUN_HEADER           m_pHeader;
+    bool                            m_isComponentAssembly;
+    NativeImage*                    m_pNativeImage;
```
E.g. These need to be `PTR_NativeImage`, etc. like all other pointers in this structure to work correctly under DAC.
